### PR TITLE
Promote live location labs flag [PSF-959]

### PR DESCRIPTION
**Pull request by onurays**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

We want to show users the Live Location Sharing option with an informative dialog.

## Screenshots / GIFs

|Screenshot|
|-|
|lls_promoting_labs_flag|

## Tests

- Disable the live location sharing labs setting if it is enabled
- In a room, choose the location sharing attachment option from the composer
- Choose "Share live location"
- You should see the dialog like the attached screenshot above
- If the user enables location sharing and clicks the OK button then the flag in labs settings should also be enabled
- Share a live location

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -3088,4 +3088,10 @@
     <string name="settings_troubleshoot_test_current_endpoint_failed">Cannot find the endpoint.</string>
     <string name="settings_troubleshoot_test_current_gateway_title">Gateway</string>
     <string name="settings_troubleshoot_test_current_gateway">Current gateway: %s</string>
+
+    <!-- Live Location Sharing Labs Flag Promotional BottomSheet -->
+    <string name="live_location_labs_promotion_title">Live location sharing</string>
+    <string name="live_location_labs_promotion_description">Please note: this is a labs feature using a temporary implementation. This means you will not be able to delete your location history, and advanced users will be able to see your location history even after you stop sharing your live location with this room.</string>
```
Comment (ouchadam) [inline vector/src/main/res/values/strings.xml:3094]: is this the final copy? it's a little bit _techy_~

File: vector/src/main/java/im/vector/app/features/location/LocationSharingFragment.kt [hunk 2]
```diff
@@ -194,6 +195,25 @@ class LocationSharingFragment @Inject constructor(
         }
     }
 
+    private val liveLocationLabsFlagPromotionListener = object : VectorBaseBottomSheetDialogFragment.ResultListener {
+        override fun onBottomSheetResult(resultCode: Int, data: Any?) {
+            // Check if the user wants to enable the labs flag
+            if (resultCode == VectorBaseBottomSheetDialogFragment.ResultListener.RESULT_OK && (data as? Boolean) == true) {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/location/LocationSharingFragment.kt:201]: this comment above could be replaced by extracting this condition to its own function

```kotlin
private fun hasUserOptedInToLiveLocationLabs(code, data) {
  return resultCode == VectorBaseBottomSheetDialogFragment.ResultListener.RESULT_OK && (data as? Boolean) == true 
}
```

**Review by ouchadam — APPROVED** [review]
tiny refactor comment, not a blocker, LGTM :100:

File: vector/src/main/res/values/strings.xml (hunk 1 above)
Comment (onurays) [reply vector/src/main/res/values/strings.xml:3094]: Yes it is. I think it is not so hard to understand for users.

File: vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt [hunk 3]
```diff
@@ -1047,6 +1047,12 @@ class VectorPreferences @Inject constructor(
         return defaultPrefs.getBoolean(SETTINGS_LABS_ENABLE_LIVE_LOCATION, false)
     }
 
+    fun setLiveLocationLabsEnabled() {
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt:1050]: add parameter?

**Review by ganfra — COMMENTED** [review]
Small remarks

File: vector/src/main/java/im/vector/app/features/location/LocationSharingFragment.kt (hunk 2 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/location/LocationSharingFragment.kt]: Right, done.

File: vector/src/main/java/im/vector/app/features/location/live/LiveLocationLabsFlagPromotionBottomSheet.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            val enableLabsFlag = views.promoteLiveLocationFlagSwitch.isChecked
+            resultListener?.onBottomSheetResult(ResultListener.RESULT_OK, enableLabsFlag)
+            dismiss()
+        }
+    }
+
+    companion object {
+        fun newInstance(resultListener: ResultListener): LiveLocationLabsFlagPromotionBottomSheet {
+            return LiveLocationLabsFlagPromotionBottomSheet().also {
+                it.resultListener = resultListener
```
[the remark this answers is not in the document]
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/location/live/LiveLocationLabsFlagPromotionBottomSheet.kt]: Nice catch. Fixed.

**Review by ganfra — CHANGES_REQUESTED** [review]
See my remark, sorry I should have detailed more in the first place.
Also fix the ktlint!

**Review by ganfra — APPROVED** [review]

**Comment by ganfra** [comment]
Lets merge this is everything is green!

**Pull request merged**